File: Napakalaki_Ruby/lib/bad_consequence.rb
# encoding: utf-8

=begin
********************************************************************
**       _   _                   _         _       _    _         **
**      | \ | | __ _ _ __   __ _| | ____ _| | __ _| | _(_)        **
**      |  \| |/ _` | '_ \ / _` | |/ / _` | |/ _` | |/ / |        **
**      | |\  | (_| | |_) | (_| |   < (_| | | (_| |   <| |        **
**      |_| \_|\__,_| .__/ \__,_|_|\_\__,_|_|\__,_|_|\_\_|        **
**                  |_|                                           **
** Creado por Javier Bolívar Valverde <javibolivar@correo.ugr.es> **
** Programación y Diseño Orientado a Objetos                      **
** Grado en Ingeniería Informática (Universidad de Granada)       **
********************************************************************
=end

class BadConsequence
  def BadConsequence::maxtreasures
    @@MAXTREASURES = 10
  end

  def initialize(t, l)
    @text = t
    @levels = l
  end
  
  private_class_method :new #Clase abstracta
  
  attr_accessor :text, :levels, :n_visible_treasures, :n_hidden_treasures, :specific_visible_treasures, :specific_hidden_treasures, :MAXTREASURES

  #Devuelve true cuando el mal rollo que tiene que cumplir el jugador está vacío, eso
  #significa que el conjunto de atributos del mal rollo indican que no hay mal rollo que
  #cumplir
  def is_empty
    raise NotImplementError.new
  end

  def substract_visible_treasure(t)
    raise NotImplementError.new
  end

  def substract_hidden_treasure(t)
    raise NotImplementError.new
  end

  def adjust_to_fit_treasure_lists(v, h)
    raise NotImplementError.new
  end

  def to_s
    malrollo = "Mal rollo: \n" + "Nombre: " + @text + "\n"
    if(@levels != 0)
      malrollo += "Niveles perdidos: " + @levels.to_s + "\n"
    end
    
    return malrollo
  end
end

File: Napakalaki_Java/src/napakalaki/BadConsequence.java
/*
********************************************************************
**       _   _                   _         _       _    _         **
**      | \ | | __ _ _ __   __ _| | ____ _| | __ _| | _(_)        **
**      |  \| |/ _` | '_ \ / _` | |/ / _` | |/ _` | |/ / |        **
**      | |\  | (_| | |_) | (_| |   < (_| | | (_| |   <| |        **
**      |_| \_|\__,_| .__/ \__,_|_|\_\__,_|_|\__,_|_|\_\_|        **
**                  |_|                                           **
** Creado por Javier Bolívar Valverde <javibolivar@correo.ugr.es> **
** Programación y Diseño Orientado a Objetos                      **
** Grado en Ingeniería Informática (Universidad de Granada)       **
********************************************************************
*/
package napakalaki;

import java.util.ArrayList;

public abstract class BadConsequence {
    //Atributos
    static final int MAXTREASURES = 10;
    
    protected String text;
    protected int levels;
    
    //Constructor
    public BadConsequence(String t, int l) {
        this.text = t;
        this.levels = l;
    }
    
    //Métodos
    public String getText() {
        return this.text;
    }
    
    public int getLevels() {
        return this.levels;
    }
    
    public int getNVisibleTreasures() {
        return 0;
    }
    
    public int getNHiddenTreasures() {
        return 0;
    }
    
    public ArrayList<TreasureKind> getSpecificVisibleTreasures() {
        return null;
    }
    
    public ArrayList<TreasureKind> getSpecificHiddenTreasures() {
        return null;
    }
    
    public abstract BadConsequence adjustToFitTreasureList(ArrayList<Treasure> v, ArrayList<Treasure> h);
    
    public abstract void substractVisibleTreasure(Treasure t);
    
    public abstract void substractHiddenTreasure(Treasure t);
    
    public abstract boolean isEmpty();
    
    @Override
    public String toString() {
        String malrollo = "Mal rollo genérico: \n" + "Nombre: " + this.text + "\n";
        if (this.levels != 0)
            malrollo += "Niveles perdidos: " + this.levels + "\n";
        
        return malrollo;
    }
}
